deployer/steps.go

```diff
@@ -52,17 +59,31 @@ func (d *Deployer) CleanTmpDirs() error {
 	return err.ErrorOrNil()
 }
 
-func (d *Deployer) StepPrepISODir() error {
-	return d.Add(constants.OpPrepareISO, herd.WithCallback(func(ctx context.Context) error {
+// StepPrepDestination prepares the destination directory for the rest of the steps.
+// This is the first step always executed in the deployer, it creates the destination directory in which other build steps will operate.
+func (d *Deployer) StepPrepDestination() error {
+	return d.Add(constants.OpPrepareDestination, herd.WithCallback(func(ctx context.Context) error {
+		internal.Log.Logger.Info().Str("destination", d.destination()).Msg("Preparing destination temporal directory")
+		if d.destination() == "" {
+			internal.Log.Logger.Error().Msg("Destination directory is not set, cannot prepare ISO directory")
+			return fmt.Errorf("destination directory is not set")
+		}
 		return os.MkdirAll(d.destination(), 0755)
 	}))
 }
 
 func (d *Deployer) StepCopyCloudConfig() error {
 	return d.Add(constants.OpCopyCloudConfig,
-		herd.WithDeps(constants.OpPrepareISO),
+		herd.WithDeps(constants.OpPrepareDestination, constants.OpPrepareNetboot, constants.OpPreparetmproot),
 		herd.WithCallback(func(ctx context.Context) error {
 			internal.Log.Logger.Info().Str("cloudConfig", d.Config.CloudConfig).Msg("Copying cloud config")
+			if _, err := os.Stat(d.destination()); err != nil && os.IsNotExist(err) {
```

Comment:
`StepPrepDestination` already does that and it a dependency of this step. Is this what you mean in the description that it has to run at the right point in time? If that's true, what is the purpose of the `StepPrepDestination` ? Maybe what we need is a `herd.Pre` kind of thing that let's us run a function before the actual step runs and we prepare the destination there (maybe it's already supported in herd? I didn't check).

Reply:
yeah I think I went overboard with extra checks around as it was such a mess. Ideally this should never happen, but that was what we expected before and it still happened. So I added this so we can see the error properly if we mess this up again somehow (we will xD)